devicetypes/heltun/heltun-tps05-switch.src/heltun-tps05-switch.groovy

```diff
@@ -0,0 +1,692 @@
+	def numberOfButtons = state.numberOfButtons
+	def value = cmd.value
+	def childDevice = childDevices?.find {channelNumber(it.deviceNetworkId) == ep }
+	def corrRelCons = 0
+	def corRelParam = 11 + ep - numberOfButtons
+	if (ep in numberOfButtons..(2*numberOfButtons)) {
+		def param = parameterMap().find( {it.paramNum == corRelParam } ).name
+		def paramState = state."$param"
+		if (paramState) {
+			corrRelCons = paramState.value
+		}
+	}
+	if (childDevice) {
+		childDevice.sendEvent(name: "switch", value: value ? "on" : "off")
+		if (value) {
+			sendEvent(name: "switch", value: "on")
+			childDevice.sendEvent(name: "power", value: corrRelCons, unit: "W")
+		} else {
+			childDevice.sendEvent(name: "power", value: 0, unit: "W")
+			if (!childDevices.any { it.currentValue("switch") == "on" }) {
+				sendEvent(name: "switch", value: "off")
+			}
+		}
+	}
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.clockv1.ClockReport cmd) {
+	def currDate = Calendar.getInstance(location.timeZone)
+	def time = [hour: currDate.get(Calendar.HOUR_OF_DAY), minute: currDate.get(Calendar.MINUTE), weekday: currDate.get(Calendar.DAY_OF_WEEK)]
+	if ((time.hour != cmd.hour) || (time.minute != cmd.minute) || (time.weekday != cmd.weekday)) {
+		sendHubCommand(zwave.clockV1.clockSet(time).format())
+	}
+}
+
+def zwaveEvent(physicalgraph.zwave.commands.multichannelassociationv2.MultiChannelAssociationReport cmd) {
+	def cmds = []
+	if (cmd.groupingIdentifier == 1) {
+		if (cmd.nodeId != [0, zwaveHubNodeId, 0]) {
+			cmds << zwave.multiChannelAssociationV2.multiChannelAssociationRemove(groupingIdentifier: 1).format()
+			cmds << zwave.multiChannelAssociationV2.multiChannelAssociationSet(groupingIdentifier: 1, nodeId: [0,zwaveHubNodeId,0]).format()
+		}
+	}
+	if (cmds) {
+		sendHubCommand(cmds, 1200)
+	}
+}
+
+def configure() {
+	refresh()
+}
+
+def refresh() {
+	def numberOfButtons = state.numberOfButtons
+	def cmds = []
+	cmds << zwave.sensorMultilevelV5.sensorMultilevelGet(sensorType:roomTemperature).format()
+	cmds << zwave.sensorMultilevelV5.sensorMultilevelGet(sensorType:humidity).format()
+	cmds << zwave.sensorMultilevelV5.sensorMultilevelGet(sensorType:illuminance).format()
+	for (i in 1..(2 * numberOfButtons)) {
+		cmds << encap(zwave.switchBinaryV1.switchBinaryGet(), i)
+	}
+	cmds << zwave.clockV1.clockGet().format()
+	cmds << zwave.multiChannelAssociationV2.multiChannelAssociationGet(groupingIdentifier: 1).format()
+	sendHubCommand(cmds, 1200)
+	runIn(15, "checkParam")
+}
+
+def ping() {
+	def numberOfButtons = state.numberOfButtons
+	def cmds = []
+	cmds << zwave.sensorMultilevelV5.sensorMultilevelGet(sensorType:roomTemperature).format()
+	cmds << zwave.sensorMultilevelV5.sensorMultilevelGet(sensorType:humidity).format()
+	cmds << zwave.sensorMultilevelV5.sensorMultilevelGet(sensorType:illuminance).format()
+	for (i in 1..(2 * numberOfButtons)) {
+		cmds << encap(zwave.switchBinaryV1.switchBinaryGet(), i)
+	}
```

Comment:
Please move it to a separate function -> for example def getRefreshCommands(), 
then call: 
`def cmds = getRefreshCommands()`
to avoid code duplication in `ping()` and in `refresh()`